tests/src/upgrade.rs

```diff
@@ -0,0 +1,401 @@
+use casper_engine_test_support::{
+    ExecuteRequestBuilder, InMemoryWasmTestBuilder, DEFAULT_ACCOUNT_ADDR,
+    DEFAULT_RUN_GENESIS_REQUEST,
+};
+use casper_types::{account::AccountHash, runtime_args, CLValue, Key, RuntimeArgs};
+
+use crate::utility::{
+    constants::{
+        ACCESS_KEY_NAME, ACCOUNT_USER_1, ARG_IS_HASH_IDENTIFIER_MODE, ARG_NFT_CONTRACT_HASH,
+        ARG_NFT_PACKAGE_HASH, ARG_SOURCE_KEY, ARG_TARGET_KEY, ARG_TOKEN_HASH, ARG_TOKEN_META_DATA,
+        ARG_TOKEN_OWNER, BACKFILLED_TOKEN_TRACKER, CONTRACT_1_0_0_WASM, MAX_PAGE_NUMBER,
+        MIGRATE_WASM, MINT_SESSION_WASM, NFT_CONTRACT_WASM, NFT_TEST_COLLECTION, NFT_TEST_SYMBOL,
+        PAGE_SIZE, RECEIPT_NAME, TOKEN_COUNT_AT_UPGRADE, TRANSFER_SESSION_WASM,
+    },
+    installer_request_builder::{
+        InstallerRequestBuilder, MetadataMutability, NFTIdentifierMode, NFTMetadataKind,
+        OwnershipMode,
+    },
+    support,
+};
+
+const OWNED_TOKENS: &str = "owned_tokens";
+
+#[test]
+fn should_safely_upgrade_in_ordinal_identifier_mode() {
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&DEFAULT_RUN_GENESIS_REQUEST).commit();
+
+    let install_request = InstallerRequestBuilder::new(*DEFAULT_ACCOUNT_ADDR, CONTRACT_1_0_0_WASM)
+        .with_collection_name(NFT_TEST_COLLECTION.to_string())
+        .with_collection_symbol(NFT_TEST_SYMBOL.to_string())
+        .with_total_token_supply(100u64)
+        .with_ownership_mode(OwnershipMode::Minter)
+        .with_identifier_mode(NFTIdentifierMode::Ordinal)
+        .with_nft_metadata_kind(NFTMetadataKind::Raw)
+        .build();
+
+    builder.exec(install_request).expect_success().commit();
+
+    let nft_contract_hash_1_0_0 = support::get_nft_contract_hash(&builder);
+    let nft_contract_key_1_0_0: Key = nft_contract_hash_1_0_0.into();
+
+    for _ in 0..3 {
```

Comment:
mention why 3 is arbitrary, just used to put some state in before upgrade

Reply:
Added comment

Reply:
Extracted to a variable